### PR TITLE
+ no_wildcard_variable_uses

### LINKED ISSUE #4392: proposal: `no_wildcard_variable_uses`
## Description

Do not use wildcard parameters or variables.

## Details

There's a proposal [to change the language](https://github.com/dart-lang/language/blob/master/working/wildcards/feature-specification.md) so that (at least) parameters and local variables named `_` (e.g., wildcards) become non-binding. When that happens, existing code that uses wildcard parameters or variables will break. 

Enabling this lint now will make it less breaking to make `_` non-binding in a future Dart release.

## Kind

Error

## Bad Examples

```dart
var _ = 1;
print(_); // LINT
```

```dart
void f(int __) {
  print(__); // LINT multiple underscores too
}
```

## Good Examples 

```dart
for (var _ in [1, 2, 3]) count++;
```

```dart
try {
  ...
} catch (_) {
  ...
}
```

```dart
var [a, _, b, _] = [1, 2, 3, 4];
```

## Discussion

https://github.com/dart-lang/sdk/issues/51221

cc @dart-lang/language-team

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

### PR DESCRIPTION
**Pull request by pq**
Fixes #4392

/cc @bwilkerson @scheglov @MaryaBelanger

### CODE REVIEW

File: lib/src/rules/no_wildcard_variable_uses.dart [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitSimpleIdentifier(SimpleIdentifier node) {
+    if (node.name.isJustUnderscores) {
```
Comment (pq) [inline lib/src/rules/no_wildcard_variable_uses.dart:75]: I *think* it's just this simple... Please tell me if I'm over-simplifying though!

File: lib/src/rules/no_wildcard_variable_uses.dart [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+import '../util/ascii_utils.dart';
+
+const _desc = r"Don't use wildcard parameters or variables.";
+
+const _details = r'''
+**DON'T** use wildcard parameters or variables.
+
```
Comment (pq) [inline lib/src/rules/no_wildcard_variable_uses.dart:15]: @MaryaBelanger: any word-smithing appreciated!

File: lib/src/rules/no_wildcard_variable_uses.dart (hunk 1 above)
Comment (scheglov) [reply lib/src/rules/no_wildcard_variable_uses.dart:75]: The lint says: don't *use*.
Why not: don't *declare*?
There might be multiple uses, but they original reason is that something declared it.

Also, this visit method will fire on getters `print(this._)`.
Do we want this?

Comment (bwilkerson) [reply lib/src/rules/no_wildcard_variable_uses.dart:75]: > Why not: don't declare?

Read https://github.com/dart-lang/sdk/issues/51221.

> Do we want this?

No, the lint should only fire for uses of local variables (including parameters). We need tests that verify that it doesn't fire for members, including constructors and top-level functions, getters, etc.

Comment (pq) [reply lib/src/rules/no_wildcard_variable_uses.dart:75]: Thanks @bwilkerson. I'd forgotten about class members.

Comment (scheglov) [reply lib/src/rules/no_wildcard_variable_uses.dart:75]: Ah, of course I should have read the issue discussion.

Comment (pq) [reply lib/src/rules/no_wildcard_variable_uses.dart:78]: OK. Added executable element filtering (and tests).  Thanks!

File: lib/src/rules/no_wildcard_variable_uses.dart [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitSimpleIdentifier(SimpleIdentifier node) {
+    if (node.staticElement is ExecutableElement) return;
```
Comment (scheglov) [inline lib/src/rules/no_wildcard_variable_uses.dart:76]: This looks too vague in my opinion.
Why to explicitly say what we want to check?
`element is LocalVariableElement || element is ParameterElement`

Comment (pq) [reply lib/src/rules/no_wildcard_variable_uses.dart]: SGTM. Updated.

**Review by scheglov — APPROVED** [review]

File: lib/src/rules/no_wildcard_variable_uses.dart [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+const _desc = r"Don't use wildcard parameters or variables.";
+
+const _details = r'''
+**DON'T** use wildcard parameters or variables.
+
+In a future version of the Dart langauge, wildcard parameters and local
+variables (e.g., named with just underscores like `_`, `__`, `___`, etc.) will
+become non-binding. When that happens, existing code that uses wildcard
+parameters or variables will break. In anticipation and to make adoption
+easier, this lint disallows wildcard and variable parameter uses.
```
Comment (MaryaBelanger) [inline lib/src/rules/no_wildcard_variable_uses.dart:21]:
```suggestion
Wildcard parameters and local variables
(e.g. underscore-only names like `_`, `__`, `___`, etc.) will
become non-binding in a future version of the Dart language.
Any existing code that uses wildcard parameters or variables will
break. In anticipation of this change, and to make adoption easier,
this lint disallows wildcard and variable parameter uses.
```

**Review by MaryaBelanger — APPROVED** [review]
LGTM! Just a slight reword, take it or leave it :)

**Pull request merged**